h/static/scripts/directive/annotation.js

```diff
@@ -0,0 +1,730 @@
+        return selector.type === 'TextQuoteSelector';
+      });
+    });
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#authorize
+    * @param {string} action The action to authorize.
+    * @returns {boolean} True if the action is authorized for the current user.
+    * @description Checks whether the current user can perform an action on
+    * the annotation.
+    */
+  vm.authorize = function(action) {
+    if (model === null) {
+      return false;
+    }
+    // TODO: this should use auth instead of permissions but we might need
+    // an auth cache or the JWT -> userid decoding might start to be a
+    // performance bottleneck and we would need to get the id token into the
+    // session, which we should probably do anyway (and move to opaque bearer
+    // tokens for the access token).
+    return permissions.permits(action, model, session.state.userid);
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#delete
+    * @description Deletes the annotation.
+    */
+  vm['delete'] = function() {
+    return $timeout(function() {  // Don't use confirm inside the digest cycle.
+      var onRejected;
+      var msg = 'Are you sure you want to delete this annotation?';
+      if ($window.confirm(msg)) {
+        onRejected = function(reason) {
+          flash.error(
+            errorMessage(reason), 'Deleting annotation failed');
+        };
+        $scope.$apply(function() {
+          annotationMapper.deleteAnnotation(model).then(
+            null, onRejected);
+        });
+      }
+    }, true);
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#edit
+    * @description Switches the view to an editor.
+    */
+  vm.edit = function() {
+    if (!drafts.get(model)) {
+      updateDraft(model);
+    }
+    vm.action = model.id !== null ? 'edit' : 'create';
+    vm.editing = true;
+    vm.preview = 'no';
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#view
+    * @description Switches the view to a viewer, closing the editor controls
+    *              if they are open.
+    */
+  vm.view = function() {
+    vm.editing = false;
+    vm.action = 'view';
+  };
+
+  /**
+    * @ngdoc method
+    * @name annotation.AnnotationController#revert
+    * @description Reverts an edit in progress and returns to the viewer.
+    */
+  vm.revert = function() {
+    drafts.remove(model);
+    if (vm.action === 'create') {
+      $rootScope.$emit('annotationDeleted', model);
+    } else {
+      vm.render();
+      vm.view();
+    }
+  };
+
+  /**
+  * Update the given annotation domain model object with the data from the
+  * given annotation view model object.
+  */
+  updateDomainModel = function(domainModel, viewModel) {
+    var i;
+    var tagTexts = [];
+    for (i = 0; i < viewModel.tags.length; i++) {
+      tagTexts.concat(viewModel.tags[i].text);
```

Comment:
Array.concat() does not mutate its argument, it returns a new array.

This could be written more simply as:

```js
var tagTexts = viewModel.tags.map(function (tag) { return tag.text });
```

Also, there is a typo in the argument to angular.extend. Should be `tags` not `tag`.